### DISCUSSION TITLE
Chaining nodes for several turn LLM conversation?

### DISCUSSION THREAD

**Discussion by Jonseed** [glibsonoran/Plush-for-ComfyUI#132] [Q&A]
How would you recommend chaining the nodes together if you want to prompt the LLM with several prompts but keep the context/history of a single conversation? Do you just take the LLMprompt output into the Examples_and_context input of the next node? Or is there a better way of multi-message conversations?

**Comment by glibsonoran** [comment]
Examples_and_context is where you would keep track of context i.e.: the conversation.  Generally this needs to be in the form:  [User Prompt; model response], repeated as many times as desired bearing in mind that when the model's context window is full it will start dropping the oldest text.  

Some Services and Models are very strict about this structure (Anthropic) and some are more forgiving; like if you provide two prompts or responses sequentially.  So you'd want to use a concatenate node to put the prompt over the response with a delimiter that's the one you've set as Advanced Prompt Enhancer's examples_delimiter ( like :: or | ), and you'd want to tell the model to end its output with that same delimiter so you could chain them.  

When I try to set this up to fill automatically ComfyUI won't accept the workflow (probably because there are circular connections).  So I place the most recent context item output next to the text field that's holding the Examples_and_context input, and paste each new context item from one to the other at the end of the queue.

I added {{user}} and {{model}} for clarity, they're not necessary.

![Context](https://github.com/user-attachments/assets/aa55e55a-a308-4e86-9edf-88fdda10501d)

**Reply by Jonseed** [reply]
I wonder if you could add a "context" output to the Advanced Prompt Enhancer node that would make this chaining process easier. Then you could just link up that context output to the example/context input of the next node.  Or would this not work since the different services take different formats? (Or maybe your node could take that into account based on which service is selected.)

The [ComfyUI-Ollama](https://github.com/stavsap/comfyui-ollama?tab=readme-ov-file#ollamagenerateadvance) nodes have this kind of setup, but of course they are only dealing with Ollama, so it's a bit easier.

![generate-advance.png](https://github.com/user-attachments/assets/b9c8a2e0-6176-4aa1-9aa6-8442314c28e5)

**Reply by glibsonoran** [reply]
Context output is almost done... I'm testing it right now :-D:
![Context_out](https://github.com/user-attachments/assets/08bb5501-5251-4a0b-9d40-35a13628ff9e)

**Reply by Jonseed** [reply]
That's awesome!

**Reply by EnragedAntelope** [reply]
> That's awesome!

I am trying to understand use case - do you have to prefill certain steps in the conversation? You will use comfyui instead of Sillytavern/etc. for role play or storytelling?

**Reply by Jonseed** [reply]
@EnragedAntelope sometimes it makes sense to prompt the LLM in separate steps, like if you are going to use its response in a subsequent query.

**Reply by EnragedAntelope** [reply]
> @EnragedAntelope sometimes it makes sense to prompt the LLM in separate steps, like if you are going to use its response in a subsequent query.

Thanks. Still trying to think up a precise use case in ComfyUI for this but my brain isn't being all that creative so far :)

**Comment by glibsonoran** [comment]
One added note: if you preface the tag in tagger with two colons: '::{{user}}' then you don't have to tell the model to end its output with two colons. 
Also, I have an Example Workflow that explains how to use Examples_and_context: `How_To_Use_Examples.png`
